File: leetcode/cpp/P0435_NonOverlappingIntervals.cpp
// https://leetcode.com/problems/non-overlapping-intervals/

#include <algorithm>
#include <vector>

class Solution {
 public:
  int eraseOverlapIntervals(std::vector<std::vector<int>>& intervals) {
    if (intervals.size() == 1) {
      return 0;
    }

    std::sort(intervals.begin(), intervals.end(),
              [](const std::vector<int>& lhs, const std::vector<int>& rhs) {
                if (lhs[0] == rhs[0]) {
                  return lhs[1] < rhs[1];
                }
                return lhs[0] < rhs[0];
              });

    int prev = 0;
    int count = 0;

    for (int ii = 1; ii < intervals.size(); ++ii) {
      if (intervals[prev][1] > intervals[ii][0]) {
        if (intervals[prev][1] > intervals[ii][1]) {
          prev = ii;
        }
        ++count;
      } else {
        prev = ii;
      }
    }

    return count;
  }
};

File: leetcode/python/P0435_NonOverlappingIntervals.py
# https://leetcode.com/problems/non-overlapping-intervals/
# Difficulty: Medium

class Solution:
    def eraseOverlapIntervals(self, intervals: list[list[int]]) -> int:
        # n = len(intervals) <= 10^5 ==> max complexity nlogn
        # sort the intervals; then, compare the current interval with the "previous"
        # one;
        #
        # choose greedily to keep the
        #
        # prev,   crt
        # [1, 2], [1, 3], [2, 3], [3, 4]
        # [1, 2]; [1, 3] --> 2 > 1 ==> overlap;
        # but 3 > 2 ==> the current interval is shorter, keep it;
        # discard intervals greedily (intuition: wider interval has
        # higher chance to overlap)

        # TODO: Read solution and proof here (from editorial):
        # https://www.cs.umd.edu/class/fall2017/cmsc451-0101/Lects/lect07-greedy-sched.pdf
        intervals.sort()
        num_remove = 0
        prev_idx = 0

        for idx in range(1, len(intervals)):
            if intervals[prev_idx][1] > intervals[idx][0]:
                # discard the
                if intervals[prev_idx][1] > intervals[idx][1]:
                    prev_idx = idx
                num_remove += 1
            else:
                prev_idx = idx
        return num_remove

if __name__ == "__main__":
    s = Solution()
    print(s.eraseOverlapIntervals([[1,2],[2,3],[3,4],[1,3]]))  # 1
    print(s.eraseOverlapIntervals([[1,2],[1,2],[1,2]]))  # 2
    print(s.eraseOverlapIntervals([[1,2],[2,3]]))  # 0
    print(s.eraseOverlapIntervals([[1,100],[11,22],[1,11],[2,12]]))  # 2
    print(s.eraseOverlapIntervals([[0,2],[1,3],[2,4],[3,5],[4,6]]))  # 2